### PR TITLE
log in component and page

### LINKED ISSUE #60: Login Form Component
# Login Form Component

- [ ] Props
- [ ] Components
- Form [Ant Design Form](https://ant.design/components/form/)
- Search Bar
- [ ] Snapshot

image

### PR DESCRIPTION
**Pull request by AlaaShurrab**
fixes #60
log in component and page
image

### CODE REVIEW

File: client/src/App/index.js [hunk 1]
```diff
@@ -1,23 +1,17 @@
...
-        <a
-          className="App-link"
-          href="https://reactjs.org"
-          target="_blank"
-          rel="noopener noreferrer"
-        >
-          Learn React
-        </a>
+        <Switch>
+          <Route exact path="/login">
```
Comment (Mu7ammadAbed) [inline client/src/App/index.js:11]:
```suggestion
          <Route exact path="/dashboard/login">
```

Comment (AlaaShurrab) [reply client/src/App/index.js:11]: but it is a free route and the dashboard is supposed to be protected?

File: client/src/components/logInForm.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      {error && <Alert message={error} type="error" showIcon />}
+      <Form.Item
+        label="Email"
+        name="email"
+        rules={[
+          {
+            required: true,
+            type: 'email',
+            message: 'Please input your email!',
+          },
```
Comment (Mu7ammadAbed) [inline client/src/components/logInForm.js:60]:
```suggestion
         {
            required: true,
            message: 'Please input your email!',
          },
          {
            type: 'email',
            message: 'Enter a valid email!',
          },
```

File: client/src/components/logInForm.js [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      </Form.Item>
+
+      <Form.Item
+        label="Password"
+        name="password"
+        rules={[
+          {
+            required: true,
+            message: 'Please input your password!',
+          },
```
Comment (Mu7ammadAbed) [inline client/src/components/logInForm.js:73]:
```suggestion
          {
            required: true,
            message: 'Please input your password!',
          },
          {
            min: 8,
            message: 'Password must be at least 8 characters.',
          },
```

File: client/src/components/logInForm.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  const onFinish = async ({ email, password }) => {
+    try {
+      await axios.post('/api/v1/login', {
+        email,
+        password,
+      });
+      setIsAuth(true);
+      history.push('/');
+    } catch (srvError) {
+      setError('Something went wrong');
```
Comment (Mu7ammadAbed) [inline client/src/components/logInForm.js:38]:
```suggestion
      setError(srvError.response.data.message || `Something went wrong`);
      setIsLoading(false);
```

File: client/src/components/logInForm.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+const tailLayout = {
+  wrapperCol: {
+    offset: 8,
+    span: 16,
+  },
+};
+
+const LoginForm = () => {
+  const [error, setError] = useState('');
```
Comment (Mu7ammadAbed) [inline client/src/components/logInForm.js:24]:
```suggestion
  const [error, setError] = useState('');
  const [isLoading, setIsLoading] = useState('');
```

File: client/src/components/logInForm.js [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+  const { setIsAuth } = useContext(IsAuthContext);
+
+  const history = useHistory();
+  // eslint-disable-next-line consistent-return
+  const onFinish = async ({ email, password }) => {
+    try {
+      await axios.post('/api/v1/login', {
```
Comment (Mu7ammadAbed) [inline client/src/components/logInForm.js:31]:
```suggestion
    try {
      setIsLoading(true);
      await axios.post('/api/v1/login', {
```

File: client/src/components/logInForm.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        email,
+        password,
+      });
+      setIsAuth(true);
```
Comment (Mu7ammadAbed) [inline client/src/components/logInForm.js:35]:
```suggestion
      setIsAuth(true);
      setIsLoading(false);
```

File: client/src/components/logInForm.js [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+            required: true,
+            message: 'Please input your password!',
+          },
+        ]}
+      >
+        <Input.Password />
+      </Form.Item>
+
+      <Form.Item {...tailLayout}>
+        <Button type="primary" htmlType="submit">
```
Comment (Mu7ammadAbed) [inline client/src/components/logInForm.js:80]:
```suggestion
        <Button type="primary" htmlType="submit" loading={isLoading}>
```

File: client/src/App/index.js [hunk 9]
```diff
@@ -1,23 +1,24 @@
-import React from 'react';
-import logo from '../logo.svg';
+import React, { useState, useMemo } from 'react';
 import './App.css';
+import { Switch, Route } from 'react-router-dom';
+import LoginPage from '../pages/logIn';
+import IsAuthContext from '../components/context/isAuthContext';
 
 function App() {
```
Comment (Mu7ammadAbed) [inline client/src/App/index.js:7]: Change this to an arrow function

File: client/src/App/index.js [hunk 10]
```diff
@@ -1,23 +1,24 @@
...
+  const [isAuth, setIsAuth] = useState(false);
+
+  const value = useMemo(() => ({ isAuth, setIsAuth }), [isAuth, setIsAuth]);
```
Comment (Mu7ammadAbed) [inline client/src/App/index.js:10]: What is the purpose of using this?
```suggestion
```

File: client/src/App/index.js [hunk 11]
```diff
@@ -1,23 +1,24 @@
...
+
   return (
     <div className="App">
       <header className="App-header">
```
Comment (Mu7ammadAbed) [inline client/src/App/index.js:14]: Don't forget to pull the new version of the main and remove these two elements
```suggestion
```

File: client/src/App/index.js [hunk 12]
```diff
@@ -1,23 +1,24 @@
...
-        </p>
-        <a
-          className="App-link"
-          href="https://reactjs.org"
-          target="_blank"
-          rel="noopener noreferrer"
-        >
-          Learn React
-        </a>
+        <IsAuthContext.Provider value={value}>
```
Comment (Mu7ammadAbed) [inline client/src/App/index.js:15]:
```suggestion
        <IsAuthContext.Provider value={{ setIsAuth }}>
```

**Review by Mu7ammadAbed — CHANGES_REQUESTED** [review]

File: client/src/App/index.js [hunk 13]
```diff
@@ -1,23 +1,24 @@
-import React from 'react';
-import logo from '../logo.svg';
+import React, { useState, useMemo } from 'react';
 import './App.css';
+import { Switch, Route } from 'react-router-dom';
+import LoginPage from '../pages/logIn';
+import IsAuthContext from '../components/context/isAuthContext';
 
 function App() {
+  const [isAuth, setIsAuth] = useState(false);
```
Comment (Mu7ammadAbed) [inline client/src/App/index.js:8]:
```suggestion
  const [isAuth, setIsAuth] = useState(false);
  
    const checkAuth = async () => {
    try {
      await axios('/api/v1/is-auth');
      setIsAuth(true);
    } catch ({ response }) {
      setIsAuth(false);
    }
  };

  useEffect(() => {
    checkAuth();
  }, []);
```

**Review by Mu7ammadAbed — CHANGES_REQUESTED** [review]

File: client/src/components/context/isAuthContext.js [hunk 14]
```diff
@@ -0,0 +1,5 @@
+import { createContext } from 'react';
+
+const IsAuthContext = createContext(false);
+
+export default IsAuthContext;
```
Comment (hassanelnajjar) [inline client/src/components/context/isAuthContext.js:5]: Hello Alaa, 
I have some suggestion to refactor the context code :-
```js
import React, { useContext, useState, createContext } from 'react';
import { element, arrayOf } from 'prop-types';

const IsAuthContext = createContext(false);

export const AuthProvider = ({ children }) => {
  const authState = useState(false);
  return (
    <IsAuthContext.Provider value={authState}>
      {children}
    </IsAuthContext.Provider>
  );
};

AuthProvider.propTypes = {
  children: arrayOf(element).isRequired,
};
export const useAuth = () => useContext(IsAuthContext);

```
By this, You will not need to import useContext or isAuthContext when you want to use it.
Just import useAuth and invoke it and you will get the full state (value , function)

File: client/src/index.js [hunk 15]
```diff
@@ -1,11 +1,14 @@
...
 
 ReactDOM.render(
   <React.StrictMode>
-    <App />
+    <Router>
+      <App />
+    </Router>
   </React.StrictMode>,
   document.getElementById('root')
 );
```
Comment (hassanelnajjar) [inline client/src/index.js:14]: If You make the context like [this](https://github.com/GSG-G9/dentoro/pull/62/files#r609413886)
Then it's good practice to import the provide here like this 👎🏻 
```js
import React from 'react';
import ReactDOM from 'react-dom';
import { BrowserRouter as Router } from 'react-router-dom';
import { AuthProvider } from './components/context/isAuthContext';
import './index.css';
import App from './App';

ReactDOM.render(
  <React.StrictMode>
    <Router>
      <AuthProvider>
        <App />
      </AuthProvider>
    </Router>
  </React.StrictMode>,
  document.getElementById('root')
);

```

File: client/src/App/index.js [hunk 16]
```diff
@@ -1,23 +1,24 @@
...
-        >
-          Learn React
-        </a>
+        <IsAuthContext.Provider value={value}>
+          <Switch>
+            <Route exact path="/login">
+              <LoginPage />
+            </Route>
+          </Switch>
+        </IsAuthContext.Provider>
```
Comment (hassanelnajjar) [inline client/src/App/index.js:21]: If You make the context like [this](https://github.com/GSG-G9/dentoro/pull/62/files#r609413886)
Then the app file will be smaller, you don't need here to import useState or useContext or isAuthContext, Just useAuth only :)
```js
import React from 'react';
import './App.css';
import { Switch, Route } from 'react-router-dom';
import LoginPage from '../pages/logIn';
import { useAuth } from '../components/context/isAuthContext';

function App() {
  const [isAuth] = useAuth();
  return (
    <div className="App">
      <header className="App-header">
        <Switch>
          <Route exact path="/login">
            <LoginPage />
          </Route>
          <Route exact path="/">
            {isAuth ? <h2>Logged</h2> : <h2>Not Logged</h2>}
          </Route>
        </Switch>
      </header>
    </div>
  );
}

export default App;

```

I make this line as an example only 
```js
          <Route exact path="/">
            {isAuth ? <h2>Logged</h2> : <h2>Not Logged</h2>}
          </Route>
```

File: client/src/components/logInForm.js [hunk 17]
```diff
@@ -0,0 +1,88 @@
+import React, { useContext, useState } from 'react';
+import { useHistory } from 'react-router-dom';
+import { Form, Input, Button, Alert } from 'antd';
+import axios from 'axios';
+import IsAuthContext from './context/isAuthContext';
```
Comment (hassanelnajjar) [inline client/src/components/logInForm.js:5]: If You make the context like [this](https://github.com/GSG-G9/dentoro/pull/62/files#r609413886)

Then here you only need to import useAuth 
```js
import { useAuth } from './context/isAuthContext';
```

and then use it like this :)
```js
  const [, setIsAuth] = useAuth();
```

File: client/src/pages/logIn/index.js [hunk 18]
```diff
@@ -0,0 +1,18 @@
...
+import { Typography } from 'antd';
+import LoginForm from '../../components/logInForm';
+import './style.css';
+
+const { Title } = Typography;
+
+export default function LoginPage() {
+  return (
+    <div className="login">
+      <div className="header" />
```
[the remark this answers is not in the document]
Comment (AlaaShurrab) [reply client/src/pages/logIn/index.js:11]: done

File: client/src/pages/logIn/index.js [hunk 19]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+import { Typography } from 'antd';
+import LoginForm from '../../components/logInForm';
+import './style.css';
+
+const { Title } = Typography;
+
+export default function LoginPage() {
```
[the remark this answers is not in the document]
Comment (AlaaShurrab) [reply client/src/pages/logIn/index.js:8]: done

File: client/src/components/logInForm.js [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+    span: 16,
+  },
+};
+
+const LoginForm = () => {
+  const [error, setError] = useState('');
+  const { setIsAuth } = useContext(IsAuthContext);
+
+  const history = useHistory();
+  // eslint-disable-next-line consistent-return
```
[the remark this answers is not in the document]
Comment (AlaaShurrab) [reply client/src/components/logInForm.js:28]: done

File: client/src/components/logInForm.js (hunk 2 above)
Comment (AlaaShurrab) [reply client/src/components/logInForm.js:60]: done

File: client/src/components/logInForm.js (hunk 3 above)
Comment (AlaaShurrab) [reply client/src/components/logInForm.js:73]: done

File: client/src/components/logInForm.js (hunk 5 above)
Comment (AlaaShurrab) [reply client/src/components/logInForm.js:24]: done

File: client/src/components/logInForm.js (hunk 4 above)
Comment (AlaaShurrab) [reply client/src/components/logInForm.js:38]: done

File: client/src/components/logInForm.js (hunk 6 above)
Comment (AlaaShurrab) [reply client/src/components/logInForm.js:31]: done

File: client/src/components/logInForm.js (hunk 7 above)
Comment (AlaaShurrab) [reply client/src/components/logInForm.js:35]: done

File: client/src/components/logInForm.js (hunk 8 above)
Comment (AlaaShurrab) [reply client/src/components/logInForm.js:80]: done

File: client/src/App/index.js (hunk 9 above)
Comment (AlaaShurrab) [reply client/src/App/index.js:7]: done

File: client/src/App/index.js (hunk 11 above)
Comment (AlaaShurrab) [reply client/src/App/index.js:14]: done

File: client/src/App/index.js (hunk 13 above)
Comment (AlaaShurrab) [reply client/src/App/index.js:8]: done

File: client/src/App/index.js (hunk 12 above)
Comment (AlaaShurrab) [reply client/src/App/index.js]: done

File: client/src/App/index.js (hunk 10 above)
Comment (AlaaShurrab) [reply client/src/App/index.js]: i was testing and forgot to remove it

File: client/src/App/index.js (hunk 12 above)
Comment (AlaaShurrab) [reply client/src/App/index.js]: done

Comment (AlaaShurrab) [reply client/src/App/index.js]: done

**Review by hassanelnajjar — DISMISSED** [review]

File: client/src/App/index.js (hunk 12 above)
Comment (AlaaShurrab) [reply client/src/App/index.js]: done

File: client/src/App/index.js (hunk 10 above)
Comment (AlaaShurrab) [reply client/src/App/index.js]: i was testing and forgot to remove it

File: client/src/pages/logIn/style.css [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+    display: flex;
+    align-items: center;
+    justify-content: center;
+    background-color: #C3EAEE80;
+
+}
+
+.loginFormStyle{
+    width: 400px;
+}
```
Comment (hassanelnajjar) [inline client/src/pages/logIn/style.css:26]: add new line

File: client/src/pages/logIn/style.css [hunk 22]
```diff
@@ -0,0 +1,26 @@
+.loginBody{
+    padding: 100;
+    margin: 0;
+    display: flex;
+    flex-flow: column;
+    justify-content: space-around;
+    align-items: center;
+    width:600px;
+    height: 350px;
+    /* border: 3px solid #C3EAEE; */
```
Comment (hassanelnajjar) [inline client/src/pages/logIn/style.css:10]: remove comments

File: client/src/components/logInForm.js [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+      setIsLoading(true);
+      await axios.post('/api/v1/login', {
+        email,
+        password,
+      });
+      setIsAuth(true);
+      setIsLoading(false);
+      history.push('/dashboard');
+    } catch (srvError) {
+      setError(srvError.response.data.message || `Something went wrong`);
```
Comment (hassanelnajjar) [inline client/src/components/logInForm.js:40]: please re-wirte this logic 👎🏻
```js
setError(srvError.response ? srvError.response.data.message || `Something went wrong`);
```

File: client/src/components/logInForm.js [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+const tailLayout = {
+  wrapperCol: {
+    offset: 8,
+    span: 16,
+  },
+};
+
+const LoginForm = () => {
+  const [error, setError] = useState('');
+  const [isLoading, setIsLoading] = useState('');
```
Comment (hassanelnajjar) [inline client/src/components/logInForm.js:25]: make the default value as true then you don't need to setIsLoading(true) at line 31

Comment (AlaaShurrab) [reply client/src/components/logInForm.js]: done

File: client/src/components/logInForm.js (hunk 23 above)
Comment (AlaaShurrab) [reply client/src/components/logInForm.js]: done

File: client/src/pages/logIn/style.css (hunk 22 above)
Comment (AlaaShurrab) [reply client/src/pages/logIn/style.css]: done

**Review by Mu7ammadAbed — APPROVED** [review]

**Review by Alaalser — APPROVED** [review]

**Pull request merged**